### DISCUSSION TITLE
Zigpy over MQTT integration for Home Assistant's ZHA and other Zigbee integrations?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#580] [Ideas]
I'm curious about zigpy and ZHA developers thoughts on the abstraction concepts used by the new Z-Wave integration for Home Assistant?

https://www.home-assistant.io/integrations/ozw/

"**_This integration allows you to utilize OpenZWave’s ozwdaemon to control a Z-Wave network over MQTT_**"

Do you think that once matured in Home Assistant a similar concept should perhaps be considered to be implemented for zigpy and ZHA?

```
[zigpy] <--> [zigpy MQTT Daemon/wrapper] <--> [MQTT Broker] <--> [zigpy over MQTT integration]  <--> [ZHA or other integrations]
```

As I understand it, developers have been working for years on totally rewritten a new Z-Wave integration for Home Assistant using the latest OpenZWave/OZW and during the Home Assistant Conference yesterday MartinHjelmare summarized its the concept and current state:

https://www.youtube.com/watch?v=xSB_MuKkgxE&t=2535s

https://community.home-assistant.io/t/update-on-the-z-wave-integration-home-assistant-dev-docs/169045/41

https://community.home-assistant.io/t/the-future-of-z-wave-in-ha-qt-openzwave/148541

https://github.com/cgarwood/python-openzwave-mqtt

https://github.com/cgarwood/homeassistant-zwave_mqtt

https://github.com/OpenZWave/qt-openzwave

In essence, its idea is to use a kind of distributed computing design model to detach the parts that control the Z-Wave network and physical adapters/dongles (we can call those the server-side) from Home Assistant so that you can even run those parts separately on another computer on your local network, removing the need for this future Home Assistant's Z-Wave integration as the client-side to locally have a direct connection to the physical Z-Wave adapters. 

Instead of using some flakey serial TCP server as a tunnelling protocol which is currently the option for ZHA/zigpy users today to access a remote Zigbee adapter on the network, the new Z-Wave integration use MQTT as an internal transport protocol, however, the interesting part I find with that is they really only use MQTT for transport communication internally and as such do not expose its use of MQTT to the end-user on the server-side or the client-side, thus making the setup and usage more user-friendly to the end-users as they do not even really need to know that the integration use MQTT. They solve that part in the new Z-Wave integrations initial config flow which will automatically setup and configuration of an MQTT Broker if not already available, all done automatically and without user interaction.

One of the main benefits of this distributed computing design model from an end-user point-of-view is to be able to install the Zigbee coordinator adapter and zigpy on another computer, and another main benefit is that restarting Home Assistant would no longer affect your Zigbee network. As the (MQTT) transport protocol abstracts an integration like ZHA from zigpy it could also allow ZHA to use multiple instances of zigpy with each of their using their own Zigbee coordinator adapter (each of then only connected to their own network of course).

One major benefit to this design for new developers of other zigpy integrations could also be easier integration of zigpy into other projects.

**Comment by Hedda** [comment]
@pipiche38 perhaps something like this concept would be a good idea to consider for a retake of your [Zigpy plugin for Domoticz](https://github.com/pipiche38/Domoticz-Zigpy) project? 

I remember that @gizmocuz did specifically ask about possible zigpy to MQTT in https://github.com/domoticz/domoticz/issues/4312 for a Domoticz integration.

**Comment by puddly** [comment]
I was considering trying something like this but with a special radio library that wraps an existing radio library:

```
[zigpy]  <-->  [mqtt "radio" library]  <==> MQTT <==> [mqtt "radio" backend]  <-->  [existing radio library]
```

Zigpy communicates with the radio library using the `request`/`mrequest`/`broadcast` methods (which accept easily-serializable data) and uses event listeners internally so it in theory is possible without a ton of effort, but I haven't had great luck with asyncio-compatible MQTT clients and I think serialization is going to be a bit of a pain for some stuff.

**Reply by Hedda** [reply]
> I haven't had great luck with asyncio-compatible MQTT clients

Looks like MartinHjelmare might be using or at least looking at asyncio-mqtt ("MQTT client with idiomatic asyncio interface")?

https://github.com/MartinHjelmare/asyncio-mqtt

forked from

https://github.com/sbtinstruments/asyncio-mqtt

**Reply by Adminiuga** [reply]
MQTT is a subscriber/publisher model, which is not quite suitable for RPC or request/response. One solution could be moving most of ZHA functionality into ZHA MQTT publisher: 

[radio] <-> [zigpy] <-> [ZHA MQTT Publisher] <-- MQTT --> [ZHA MQTT Subscriber HA integration]

**Comment by dmulcahey** [comment]
Zigbee2MQTT already does this... I’m not too interested in this solution tbh but I am willing to watch where this goes.

**Reply by Adminiuga** [reply]
So, you are saying PRs are welcome? :)

**Reply by dmulcahey** [reply]
Not to change the existing integration... if someone wants to make another one go for it...

**Reply by Hedda** [reply]
I do not think that the new Z-Wave integration for Home Assistant uses the same concept as Zigbee2MQTT, nor would the idea be to make a new Zigbee2MQTT type of a stand-alone project based on zigpy and only have ZHA connect to it. I for one love how user-friendly that the initial setup of ZHA integration for Home Assistant has become and would not want to compromise that user experience it now has via its ease of use UI.

In the case of Zigbee2MQTT it output MQTT and that is it (it has no custom made interaction for Home Assistant which act as the client-side), while in the case of new Z-Wave integration for Home Asssiattn they focus on making an end-to-end solution which just to happen to use MQTT as their bridge protocol between the server-side and the client-side. As I understand it their goal is still focused on making the Z-Wave integration UI the focal point of that project.

I believe that the essence of the idea of that new Z-Wave integration comes back to just to use some kind of protocol to abstract the OpenZWave library (which for this discussion can be seen as approximately equal to zigpy) in order to achieve a separation of it and physical adapter from the integration component in Home Assistant. That is, "MQTT" is not the key there as they could have chosen used an other protocol, and in any case, other than mentioning it in the introduction of the project they do not really expose the fact that they use MQTT to the end-user as Zigbee2MQTT does, as I undertand that the new Z-Wave integration in reality only uses MQTT internally as a protocol between server-side and the client-side. (Perhaps [a poor choice of them to even mention "MQTT" in the new Z-Wave integration for Home Assistant](https://www.home-assistant.io/integrations/ozw/)?)

Keeping in mind that the Zigbee radios that zigpy supports rely on serial communication that is currently sensitive to packet loss or loss of sync (of serial packets) which means that only serial stream tunnelling using a simple TCP server should be to be used when wanting a stable environment. So the point to work around this could be to try to achieve some kind if bridge functionality that can also be spanned across an IP network using some kind of protocol that is resilient to packet loss and loss of sync, such as an example a Messaging Protocol or Remote Procedure Call (RPC). 

As such you would not even need to use MQTT, maybe some other protocols would be better suited?

**Reply by MattWestb** [reply]
I finding it interesting one client - server solution then one or more Zigpy is communicating with one ZHA instance and perhaps more subsystems that is "talking zigbee" and not only Zigpy. 
So the communication is being done with standard zigbee commands and ZHA is knowing in witch pool the device is hosted and can also "routing" traffic / commands between the zigbee hosts.

I dont knowing if MQTT is one good protocol for that with latency, package integrity resending and so on. What i knowing ESP-Home have doing one own API to HA  that is working well and is must faster and lesser over head then MQTT (its possible using both but only on at the time).

How is the TCP/IP implantation that is done by Zgbee Alliance working ?
Is it possible having one or more ZNCP with ethernet talking with one system like ZHA ??

**Reply by Hedda** [reply]
marcelveldt confirmed that MQTT is only used as a transport (communication protocol) internally by new solution for Z-Wave:

> You might want to read this:
> https://community.home-assistant.io/t/update-on-the-z-wave-integration-home-assistant-dev-docs/169045/41
> 
> MQTT is only used as transport (communication protocol).

This design both have pros and cons, but the point is still "only" to detach OZW and its physicial adapters from Home Assistant.

**Reply by Hedda** [reply]
> Is it possible having one or more ZNCP with ethernet talking with one system like ZHA ?

Asked before if that were possible without converting to another protocol first and one of the suggested answers was that a fast way to achieve it would to make ZHA support multiple zigpy radios, however, it would practically be like running multiple instances of ZHA but combining all devices the same UI. So each Zigbee radio would still be treated totally separately from each other and all would have their own separate Zigbee networks that would, in reality, have nothing to do with each other.

As discussed here, one way is to translate/convert Zigbee to another protocol (like example MQTT) on the server-side and then you could have the option on the client-side to implement extended support for connections to multiple servers. They mentioned that this could maybe be possible in the future for the new Z-Wave integration.

**Comment by Hedda** [comment]
> As such you would not even need to use MQTT, maybe some other protocols would be better suited?

Equidamoid mentioned earlier this year in https://github.com/zigpy/zigpy/issues/319 and https://github.com/zigpy/zigpy/issues/297 that he did not like the idea of MQTT for this purpose and that he would like to use some kind of a low-level Zigbee to RPC bridge, and as such he had previously been experimenting with dbus.

> No not that protocol. I wasn't sure about protocol at the moment.
> For the experiment it was dbus. It actually worked(*) and had all the perks of "proper" rpc, like "method call" concept or built-in introspection, so you can take a thirdparty tool (such as d-feet or qdbusviewer) and manipulate devices as "objects".
> 
> MQTT is not really made for such "call-response" way of working and the common way of abusing it for such purpose feels a but... workaround'ey to me.
> 
> For network support there is already a "zigbee encapsulation protocol". No need to do even that via mqtt.
> 
> (*) the dbus-zigbee part worked fine, I was able to call things like "OnWithTimedOff", get device properties, etc. The network itself misbehaved (seems to be fixed recently for me).

> Let me clarify my position here a little bit.
> 
> What I want to achieve is to be able to call zigbee methods (like "on with timed off" that I need the most) via some kind of RPC. Without them being hidden behind various abstraction layers.
> 
> What does exist now:
> The closest thing is zigbee2mqtt (currently struggling to get it running on my old raspbian). Not sure how stable the network will be yet, but already see 2 downsides of it:
> 
> * nodejs
> * only one zigbee dongle supported
> 
> zigpy: it's a library, so the only "client" of it has to care about everything: automations, device management, gui, etc. And right now only one such client is homeassistant, so
> 
> * writing complex automations is a bit of a pain,
> * configuring the gui requires restart of everything and AFAIR some end devices really don't appreciate when the coordinator node stops responding
> * debugging is a bit harder due to overall comlplexity
> 
> Next, deCONZ. Only provides a "high level" api. Kind of works, but sometimes loses devices in my setup. Also, bound to only one kind of hardware.
> 
> The protocol.
> I can definitely say that using this `bridge-rpc` in this context is not the best idea IMO. It's better to use one of the widely used protocols that already exist.
> 
> Even json over mqtt will be okay-ish.

**Reply by Equidamoid** [reply]
I'm not strictly against MQTT. It just doesn't provide(*) a way to have "request-response"  API which is logical to have, for example, for property requests.

(*) Yes, one can implement their own RPC over MQTT. For example, just send json-rpc messages as payload. 
It will work, but why invent yet another complex protocol instead of using something that is already there?

**Comment by Hedda** [comment]
Asking if some developer of the new Z-Wave integration for Home Assistant would maybe perhaps not terribly mind please explaining how the concept of the separate itself from how [Zigbee2MQTT](https://www.zigbee2mqtt.io/) integrates with Home Assistant and how those integrations differ from each other in regards to the development architecture and from the end-user experience/interactions?

This question is directly in regards to the follow-up question, which is, do you think that some similar concept might also be a good idea for the [ZHA integration](https://www.home-assistant.io/integrations/zha/) in Home Assistant to separate zigpy (ZHA's zigbee stack dependency) and the physical adapters/dongles?

That is, from an end-user point-of-view there are a lot of similarities between Z-Wave and Zigbee technology and both their use of physical adapters/dongles, so from a developers perspective would it not make sense if those integrations used somewhat similar architecture concepts too?

**Reply by marcelveldt** [reply]
You might want to read this:
https://community.home-assistant.io/t/update-on-the-z-wave-integration-home-assistant-dev-docs/169045/40

MQTT is only used as transport (communication protocol).

**Comment by dmulcahey** [comment]
@Hedda this was understood from the beginning. Again, I have 0 interest in this for ZHA. Z2M already works like this and replicating Z2M is not something I’m going to participate in. Also, tagging large blocks of folks is generally frowned upon.

**Reply by Hedda** [reply]
I accept your position, though please understand that for end-users Zigbee2MQTT does not work like the new Z-Wave integration for HA. End-users will in practice not need to know that the new Z-Wave solution use MQTT, and users do not have to maintain the MQTT Broker.

The new Z-Wave solution for Home Assistant has its own custom integration which makes it much more similar to the user experience to that of how ZHA works today when comparing to how Zigbee2MQTT works as its end-users still need to install [the standard MQTT integration for Home Assistant](https://www.home-assistant.io/integrations/mqtt/), and to interact with it makes it Zigbee2MQTT more similar to a third-party gateway where all components still need to maintained separately by the end-users.

**Comment by dmulcahey** [comment]
The only difference is UI work in HA.

**Reply by Hedda** [reply]
Do you have any alternative suggestions for achieving a stable and resilient connection ZHA integration in Home Assistant to a Zigbee coordinator adapter/dongle that is connected to a remote computer on your local network that is separate from your Home Assistant OS computer?

Basic serial to network proxy and serial port console server/gateway/bridge solutions like example Ser2Net or the [Tasmota serial TCP server for Sonoff ZBBridge by ITead](https://www.digiblur.com/2020/07/how-to-use-sonoff-zigbee-bridge-with.html) are not stable options for all Zigbee radios as they are sensitive to serial packet loss or loss of sync (see https://github.com/zigpy/bellows/issues/377).

**Reply by MattWestb** [reply]
If having network problems with WiFi or Ethernet its not helping having one super duper software protocol its simply not working only can perhaps making it working little better or making it worse.

Also i with Zigbee networks interferences can killing al radio communications and eliminating the interference is the only way getting it working.

Do you have trying listening on music in one bluetooth headset and making popcorn in the micro ?
Very likely you is getting one complete jamming of the 2.4Ghz in the near the micro.

**Comment by balloob** [comment]
It's super rude to tag random people @Hedda. Pure spam. 

I suggest we close this discussion.

**Reply by Hedda** [reply]
I apologise and removed the tags, however, I would argue that suggesting to close this whole discussion as spam is very rude too.

**Reply by Adminiuga** [reply]
This is not the 1st time you've been asked to stop tagging people. Removing the tags does not help.

**Comment by Adminiuga** [comment]
It is not feasible to modify ZHA to support MQTT, so it has to be a new integration, like zha-mqtt. When you have one, feel free to submit PR
